presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java

```diff
@@ -93,10 +95,18 @@ public void startWaitingForResources()
 
     private void waitForMinimumWorkers()
     {
-        ListenableFuture<?> minimumWorkerFuture = clusterSizeMonitor.waitForMinimumWorkers();
-        // when worker requirement is met, wait for query execution to finish construction and then start the execution
-        addSuccessCallback(minimumWorkerFuture, () -> addSuccessCallback(queryExecutionFuture, this::startExecution));
-        addExceptionCallback(minimumWorkerFuture, throwable -> queryExecutor.execute(() -> stateMachine.transitionToFailed(throwable)));
+        // wait for query execution to finish construction
+        addSuccessCallback(queryExecutionFuture, queryExecution -> {
+            Session session = stateMachine.getSession();
+            int executionMinCount = 1; // always wait for 1 node to be up
+            if (queryExecution.shouldWaitForMinWorkers()) {
+                executionMinCount = getRequiredWorkers(session);
+            }
+            ListenableFuture<?> minimumWorkerFuture = clusterSizeMonitor.waitForMinimumWorkers(executionMinCount, getRequiredWorkersMaxWait(session));
+            // when worker requirement is met, start the execution
+            addSuccessCallback(minimumWorkerFuture, () -> startExecution(queryExecution));
+            addExceptionCallback(minimumWorkerFuture, throwable -> queryExecutor.execute(() -> stateMachine.transitionToFailed(throwable)));
+        });
```

Comment:
@dain I noticed that `minimumWorkerFuture` is not cancelled if user cancels the query, ClusterSizeMonitor keeps reporting requirement of the cancelled query.
Would it be correct to add below code here to handle that ?
```
stateMachine.addStateChangeListener(state -> {
    if (state.isDone()) {
        minimumWorkerFuture.cancel(true);
    }
});
```

Reply:
Yes.  The current PR would rely on the timeout, but since that is user settable, we should have a way to cleanup if the query fails for some reason.   

It doesn't look like there is code to remove a canceled `minWorkerFuture` from the `futures` list.  Maybe add a listener before adding to the queue, here https://github.com/prestosql/presto/pull/2484/files#diff-e3ac2786dd162579f58ae9f8c7c644ccR108

Reply:
I think the code here https://github.com/prestosql/presto/pull/2484/files#diff-e3ac2786dd162579f58ae9f8c7c644ccR122 makes sure that `minWorkerFuture` is removed from `futuresQueue` in case it is cancelled.
I've checked that it gets called on query cancellation once I add the listener on query state here.

Reply:
Oh yes you are correct